File: app/routes/clans.py
from fastapi import APIRouter, Path, Query

from app.models import (
    CapitalRaidSeasonsResponse,
    Clan,
    ClanMembersResponse,
    ClanSearchResponse,
    ClanWar,
    ClanWarLeagueGroup,
    ClanWarLogResponse,
)
from app.routes.common import (
    STANDARD_ERROR_RESPONSES,
    pick_variant,
    respond_from_fixture,
    unsupported_variant_response,
    validate_paging,
    validate_tag,
    variant_offset,
)


router = APIRouter()


@router.get("/clans/{clanTag}/currentwar/leaguegroup", tags=["clans"], response_model=ClanWarLeagueGroup, responses=STANDARD_ERROR_RESPONSES)
async def get_clan_war_league_group(clanTag: str = Path(..., example="#2PP")):
    normalized_tag, error = validate_tag(clanTag)
    if error:
        return error
    fixture_name = pick_variant(normalized_tag, ["INWAR", "ENDED", "NOTFOUND"])
    if fixture_name is None:
        return unsupported_variant_response(normalized_tag)
    return respond_from_fixture(f"clans/leaguegroup/{fixture_name}.json")


@router.get("/clanwarleagues/wars/{warTag}", tags=["clans"], response_model=ClanWar, responses=STANDARD_ERROR_RESPONSES)
async def get_clan_war_league_war(warTag: str = Path(..., example="#2PP")):
    normalized_tag, error = validate_tag(warTag)
    if error:
        return error
    fixture_name = pick_variant(normalized_tag, ["INWAR", "PREPARATION", "ENDED"])
    if fixture_name is None:
        return unsupported_variant_response(normalized_tag)
    return respond_from_fixture(f"clans/clanwarleagues/{fixture_name}.json")


@router.get("/clans/{clanTag}/warlog", tags=["clans"], response_model=ClanWarLogResponse, responses=STANDARD_ERROR_RESPONSES)
async def get_clan_war_log(clanTag: str = Path(..., example="#2PP"), limit: int | None = Query(default=None, ge=1), after: str | None = None, before: str | None = None):
    paging_error = validate_paging(after, before)
    if paging_error:
        return paging_error
    normalized_tag, error = validate_tag(clanTag)
    if error:
        return error
    fixture_name = pick_variant(normalized_tag, ["WARLOG", "ACCESSDENIED"])
    if fixture_name is None:
        return unsupported_variant_response(normalized_tag)
    return respond_from_fixture(f"clans/warlog/{fixture_name}.json", limit=limit, after=after, before=before)


@router.get("/clans", tags=["clans"], response_model=ClanSearchResponse, responses=STANDARD_ERROR_RESPONSES)
async def search_clans(
    name: str | None = None,
    warFrequency: str | None = None,
    locationId: int | None = None,
    minMembers: int | None = None,
    maxMembers: int | None = None,
    minClanPoints: int | None = None,
    minClanLevel: int | None = None,
    limit: int | None = Query(default=None, ge=1),
    after: str | None = None,
    before: str | None = None,
    labelIds: str | None = None,
):
    paging_error = validate_paging(after, before)
    if paging_error:
        return paging_error
    return respond_from_fixture("clans/search/CLANS_FOUND.json", limit=limit, after=after, before=before)


@router.get("/clans/{clanTag}/currentwar", tags=["clans"], response_model=ClanWar, responses=STANDARD_ERROR_RESPONSES)
async def get_current_war(clanTag: str = Path(..., example="#2PP")):
    normalized_tag, error = validate_tag(clanTag)
    if error:
        return error
    fixture_name = pick_variant(normalized_tag, ["INWAR", "PREPARATION", "ENDED", "NOTINWAR", "ACCESSDENIED"])
    if fixture_name is None:
        return unsupported_variant_response(normalized_tag)
    return respond_from_fixture(f"clans/currentwar/{fixture_name}.json")


@router.get("/clans/{clanTag}", tags=["clans"], response_model=Clan, responses=STANDARD_ERROR_RESPONSES)
async def get_clan(clanTag: str = Path(..., example="#2PP")):
    normalized_tag, error = validate_tag(clanTag)
    if error:
        return error
    fixture_name = pick_variant(normalized_tag, ["CLAN", "CLAN_NO_CAPITAL_DISTRICTS"])
    if fixture_name is None:
        return unsupported_variant_response(normalized_tag)
    return respond_from_fixture(f"clans/clans/{fixture_name}.json")


@router.get("/clans/{clanTag}/members", tags=["clans"], response_model=ClanMembersResponse, responses=STANDARD_ERROR_RESPONSES)
async def get_clan_members(clanTag: str = Path(..., example="#2PP"), limit: int | None = Query(default=None, ge=1), after: str | None = None, before: str | None = None):
    paging_error = validate_paging(after, before)
    if paging_error:
        return paging_error
    normalized_tag, error = validate_tag(clanTag)
    if error:
        return error
    if variant_offset(normalized_tag) > 0:
        return unsupported_variant_response(normalized_tag)
    return respond_from_fixture("clans/members/MEMBERS.json", limit=limit, after=after, before=before)


@router.get("/clans/{clanTag}/capitalraidseasons", tags=["clans"], response_model=CapitalRaidSeasonsResponse, responses=STANDARD_ERROR_RESPONSES)
async def get_capital_raid_seasons(clanTag: str = Path(..., example="#2PP"), limit: int | None = Query(default=None, ge=1), after: str | None = None, before: str | None = None):
    paging_error = validate_paging(after, before)
    if paging_error:
        return paging_error
    normalized_tag, error = validate_tag(clanTag)
    if error:
        return error
    if variant_offset(normalized_tag) > 0:
        return unsupported_variant_response(normalized_tag)
    return respond_from_fixture("clans/capitalraidseasons/CAPITALRAIDSEASON.json", limit=limit, after=after, before=before)

File: src/handlers/clans.ts
import { HttpApiBuilder } from 'effect/unstable/httpapi';
import * as W from '../../packages/clash-contract/src/effect.js';
import { Api } from '../api/api.js';
import {
  checkPaging,
  fail,
  reply,
  run,
  variant,
} from '../fixtures/repository.js';

export const ClansLive = HttpApiBuilder.group(Api, 'clans', (handlers) =>
  handlers
    .handle('search', ({ query }) =>
      run(() => {
        if (
          query.minMembers !== undefined &&
          query.maxMembers !== undefined &&
          query.minMembers > query.maxMembers
        )
          fail(400, 'badRequest', 'minMembers must not exceed maxMembers.');
        const labels = query.labelIds
          ?.split(',')
          .map((x) => x.trim())
          .filter(Boolean);
        if (labels?.some((x) => !/^\d+$/.test(x)))
          fail(
            400,
            'badRequest',
            'labelIds must be comma-separated numeric IDs.',
          );
        return reply(
          'clans/search/CLANS_FOUND.json',
          W.ClanSearchResponse,
          query,
          (body) => ({
            ...body,
            items: body.items.filter(
              (c) =>
                (query.name === undefined ||
                  c.name
                    .toLocaleLowerCase()
                    .includes(query.name.toLocaleLowerCase())) &&
                (!query.warFrequency ||
                  query.warFrequency === 'any' ||
                  c.warFrequency === query.warFrequency) &&
                (query.locationId === undefined ||
                  c.location?.id === query.locationId) &&
                (query.minMembers === undefined ||
                  c.members >= query.minMembers) &&
                (query.maxMembers === undefined ||
                  c.members <= query.maxMembers) &&
                (query.minClanPoints === undefined ||
                  c.clanPoints >= query.minClanPoints) &&
                (query.minClanLevel === undefined ||
                  c.clanLevel >= query.minClanLevel) &&
                (!labels?.length ||
                  labels.every((id) =>
                    c.labels?.some((label) => String(label.id) === id),
                  )),
            ),
          }),
        );
      }),
    )
    .handle('profile', ({ params }) =>
      run(() =>
        reply(
          `clans/clans/${variant(params.clanTag, ['CLAN', 'CLAN_NO_CAPITAL_DISTRICTS'])}.json`,
          W.Clan,
        ),
      ),
    )
    .handle('members', ({ params, query }) =>
      run(() => {
        checkPaging(query);
        variant(params.clanTag, ['MEMBERS']);
        return reply(
          'clans/members/MEMBERS.json',
          W.ClanMembersResponse,
          query,
        );
      }),
    )
    .handle('war', ({ params }) =>
      run(() =>
        reply(
          `clans/currentwar/${variant(params.clanTag, ['INWAR', 'PREPARATION', 'ENDED', 'NOTINWAR', 'ACCESSDENIED'])}.json`,
          W.ClanWar,
        ),
      ),
    )
    .handle('warlog', ({ params, query }) =>
      run(() => {
        checkPaging(query);
        return reply(
          `clans/warlog/${variant(params.clanTag, ['WARLOG', 'ACCESSDENIED'])}.json`,
          W.ClanWarLogResponse,
          query,
        );
      }),
    )
    .handle('cwl', ({ params }) =>
      run(() =>
        reply(
          `clans/leaguegroup/${variant(params.clanTag, ['INWAR', 'ENDED', 'NOTFOUND'])}.json`,
          W.ClanWarLeagueGroup,
        ),
      ),
    )
    .handle('leagueWar', ({ params }) =>
      run(() =>
        reply(
          `clans/clanwarleagues/${variant(params.warTag, ['INWAR', 'PREPARATION', 'ENDED'])}.json`,
          W.ClanWar,
        ),
      ),
    )
    .handle('raids', ({ params, query }) =>
      run(() => {
        checkPaging(query);
        variant(params.clanTag, ['RAIDS']);
        return reply(
          'clans/capitalraidseasons/CAPITALRAIDSEASON.json',
          W.CapitalRaidSeasonsResponse,
          query,
        );
      }),
    ),
);
